packages/server/postgres/migrations/postDeploy/1625124096757_addUserEmailUniqueConstraint.ts

```diff
@@ -0,0 +1,21 @@
+import {MigrationBuilder, ColumnDefinitions} from 'node-pg-migrate'
+
+export const shorthands: ColumnDefinitions | undefined = undefined
+
+export async function up(pgm: MigrationBuilder): Promise<void> {
+  pgm.createIndex('User', ['email'], {
+    name: 'User_email_index',
+    unique: true,
+    concurrently: true
```

Comment:
concurrent indexes scare me a bit in the off chance they go invalid (e.g. a dupe email sneaks in while the index is building)

just outta curiosity did you happen to benchmark the build time of concurrent vs. locking for the User table? it'd be neat to see the difference for ~150k rows

Reply:
yes i was thinking it would be ideal if we could run the migration in an off-peak time. in that scenario, it seems the recommended course of action is to delete the invalid index and try to build the index concurrently again.

i really like the benchmarking idea. i'm biased towards creating the index concurrently, so we build the knowledge in case we have to do it again in the future with more data.

Reply:
1. I tested that with `concurrently: true` the migration would actually fail and the error reported was `CREATE INDEX CONCURRENTLY cannot run inside a transaction block`. I couldn't find a way around it.
2. I removed the `concurrently: true` option and it run just fine.
3. I did a performance test by creating a test table with 150K rows that looks like this:

<img width="322" alt="Screen Shot 2021-07-08 at 4 42 46 PM" src="https://user-images.githubusercontent.com/1879975/124891465-98568080-e00b-11eb-80b0-7f443e25f41d.png">

and then I run a similar migration script:
```javascript
  pgm.createIndex('Test', 'email', {
    name: 'Test_email_index',
    unique: true
  })
  pgm.createConstraint('Test', 'Test_email_key', 'UNIQUE USING INDEX "Test_email_index"')
```
and it didn't take too long:
```
> time yarn postmigrate 
yarn run v1.22.5
$ node scripts/postDeployMigrate.js
> Migrating files:
> - 1625124096757_addUserEmailUniqueConstraint
> - 1625733417640_testCreateIndexNonCurrently
### MIGRATION 1625124096757_addUserEmailUniqueConstraint (UP) ###
### MIGRATION 1625733417640_testCreateIndexNonCurrently (UP) ###
✨  Done in 1.88s.
yarn postmigrate  1.19s user 0.28s system 70% cpu 2.067 total
```

Let me know what you guys think!

Reply:
careful! call time !== index creation time.
the index is still building after the call is issued.
 
we'd have to query the index like
```sql
SELECT 
  now()::TIME(0), 
  a.query, 
  p.phase, 
  p.blocks_total, 
  p.blocks_done, 
  p.tuples_total, 
  p.tuples_done
FROM pg_stat_progress_create_index p 
JOIN pg_stat_activity a ON p.pid = a.pid;
```
see: https://www.depesz.com/2019/04/18/waiting-for-postgresql-12-report-progress-of-create-index-operations/

that said, i can't imagine it's more than a minute or 2, so i'm fine with moving forward.

for future reference: the node-pg driver wraps every call inside a transaction by default, so if we do want to build a concurrent index we'd just have to use the `client` strategy and manage the connection ourselves.
if we did that, we'd still have to check for an invalid index because we're deploying this same code to more than just the cloud offering, so we've gotta be sure it runs correctly on private instances, too.

```sql
SELECT * FROM pg_class, pg_index WHERE pg_index.indisvalid = false AND pg_index.indexrelid = pg_class.oid;
```